src/UI/Button.elm

```diff
@@ -32,283 +31,502 @@ import Element.Border as Border
+import UI.Internal.Basics exposing (lazyMap, maybeToList, pairUncurry, prependMaybe)
 import UI.Internal.Palette as Palette
 import UI.Internal.Primitives as Primitives
 import UI.Internal.Size as Size exposing (Size)
 import UI.Internal.Text as Text exposing (TextColor)
 import UI.Link as Link exposing (Link)
-import UI.Palette as Palette exposing (brightnessDarkest, brightnessLight, brightnessLighter, brightnessLightest, brightnessMiddle)
+import UI.Palette as Palette exposing (brightnessDarkest, brightnessLight, brightnessLighter, brightnessLightest, brightnessMiddle, tonePrimary)
 import UI.RenderConfig exposing (RenderConfig)
 import UI.Text as Text
 import UI.Utils.ARIA as ARIA
 import UI.Utils.Element as Element
 
 
 type alias Options =
-    { mode : ButtonMode
-    , tone : ButtonTone
-    , width : ButtonWidth
+    { width : ButtonWidth
     , size : Size
     }
 
 
 type alias Properties msg =
     { body : ButtonBody
-    , click : ButtonClick msg
+    , mode : ButtonMode msg
     }
 
 
 type Button msg
     = Button (Properties msg) Options
+    | Toggle (ToggleProperties msg) Options
 
 
 type ButtonBody
     = BodyText String
     | BodyIcon Icon
 
 
-type ButtonClick msg
-    = ClickMsg msg
-    | ClickLink Link
-    | ClickToggle (Bool -> msg) Bool
+type ButtonAction msg
+    = ActionMsg msg
+    | ActionRedirect Link
 
 
-type ButtonMode
-    = ModeEnabled
-    | ModeDisabled
+type ButtonMode msg
+    = ButtonWorking (ButtonAction msg) ButtonStyle
+    | ButtonDisabled
+    | ButtonSuccess
 
 
-type ButtonTone
-    = ToneDanger
+type EmbossedTone
+    = TonePrimary
+    | ToneDanger
     | ToneLight
-    | TonePrimary
-    | ToneSuccess
     | ToneClear
 
 
+type ButtonStyle
+    = StyleEmbossed EmbossedTone
+    | StyleHyperlink
+
+
 type ButtonWidth
     = WidthFull
-    | WidthRelative
+    | WidthShrink
+
+
+type alias ToggleProperties msg =
+    { current : Bool
+    , toggleMsg : Bool -> msg
+    , hint : String
+    }
 
 
 
--- Create
+-- Init
 
 
 defaultOptions : Options
 defaultOptions =
-    { mode = ModeEnabled
-    , tone = TonePrimary
-    , width = WidthRelative
+    { width = WidthShrink
     , size = Size.default
     }
 
 
-buttonAny : ButtonClick msg -> ButtonBody -> Button msg
-buttonAny click body =
-    Button (Properties body click) defaultOptions
+toggle : String -> (Bool -> msg) -> Bool -> Button msg
+toggle hint msg isEnabled =
+    Toggle { toggleMsg = msg, current = isEnabled, hint = hint } defaultOptions
 
 
-button : msg -> ButtonBody -> Button msg
-button msg =
-    buttonAny (ClickMsg msg)
+fromLabel : String -> ButtonBody
+fromLabel label =
+    BodyText label
 
 
-toggle : (Bool -> msg) -> Bool -> ButtonBody -> Button msg
-toggle msg isEnabled =
-    buttonAny (ClickToggle msg isEnabled)
+fromIcon : Icon -> ButtonBody
+fromIcon icon =
+    BodyIcon icon
 
 
-link : Link -> ButtonBody -> Button msg
-link realLink =
-    buttonAny (ClickLink realLink)
 
+-- Body to Button
```

Comment:
I will call this type of functions in a different way. Maybe entrypoints, create/init functions or something like that as this concept can be applied across all our components, would be nice to have one term to that type of function

Reply:
I found this hard: We have only one simple init "toggle" that doesn't need a body cause it's always the same: Then, we have some action-based init "cmd", and "redirect", and some static init "disabled" and "success". These last three require a body.

It's hard to fit them all in one term when what they do happens in different ways.

Also "init" is a term for system initializations, like `SDL_init(..)`, `OGLinit(..)`, not data that later may be duplicated and became four-five buttons in a list. (Considering this duplication we could also discard "new" and "create")

Other options I could think of are "define", "assemble", "begin".

Reply:
but it's not about requiring one value or another, is about being the start point of that builder.

In that cases you mentioned, all of them are the entry point, it means that they receive a bunch of things and return a Button, then that's different of the modifiers (with* functions). So in that case, `toggle`, `cmd` and `redirect` are from the same category, they are the functions used to get a type `Button`

Reply:
Done